cmd/cluster-agent/admissioncontroller/listener.go

```diff
@@ -0,0 +1,203 @@
+			body = data
+		}
+	}
+
+	log.Debug("admission controller request body: %v", body)
+
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		responseAdmissionReview.Response = newAdmissionResponseWithError(err)
+	} else {
+		if requestedAdmissionReview.Request == nil {
+			w.WriteHeader(http.StatusBadRequest)
+			log.Info("received empty request")
+			return
+		}
+
+		responseAdmissionReview.Response = handleAdmissionReview(requestedAdmissionReview)
+	}
+
+	// Return the same UID
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+
+	log.Debug("admission controller response: %v", responseAdmissionReview.Response)
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Error(err)
+		return
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Error(err)
+		return
+	}
+}
+
+// Returns a simple response with the provided error.
+// The webhook is still considered as allowed, in order
+// to not interfere with the user's deployment.
+func newAdmissionResponseWithMessage(message string, params ...interface{}) *v1beta1.AdmissionResponse {
+	msg := fmt.Sprintf(message, params...)
+	log.Error(msg)
+	return &v1beta1.AdmissionResponse{
+		Allowed: true,
+		Result: &metav1.Status{
+			Message: msg,
+		},
+	}
+}
+
+func newAdmissionResponseWithError(err error) *v1beta1.AdmissionResponse {
+	return newAdmissionResponseWithMessage(err.Error())
+}
+
+var (
+	podResource = metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
+)
+
+func handleAdmissionReview(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	if ar.Request == nil {
+		return newAdmissionResponseWithMessage("empty resource %v not supported", ar)
+	}
+
+	switch {
+	case ar.Request.Resource == podResource:
+		return handlePodRequest(ar.Request.Object.Raw)
+	default:
+		return newAdmissionResponseWithMessage("resource %v not supported", ar.Request.Resource)
+	}
+}
+
+func handlePodRequest(raw []byte) *v1beta1.AdmissionResponse {
+	pod := corev1.Pod{}
+	if err := json.Unmarshal(raw, &pod); err != nil {
+		return newAdmissionResponseWithError(err)
+	}
+
+	patch := mutatePod(pod, getEnvMutator())
+	return newJSONPatchResponse(patch)
+}
+
+var (
+	patchTypeJSONPatch = v1beta1.PatchTypeJSONPatch
+)
+
+func newJSONPatchResponse(patch jsonpatch.Patch) *v1beta1.AdmissionResponse {
+	bytes, err := json.Marshal(patch)
+	if err != nil {
+		return newAdmissionResponseWithError(err)
+	}
+
+	return &v1beta1.AdmissionResponse{
+		Allowed:   true,
+		Patch:     bytes,
+		PatchType: &patchTypeJSONPatch,
+	}
+}
+
+// NewEnvMutator creates a new mutator which adds environment
+// variables to pods
+func getEnvMutator() []corev1.EnvVar {
```

Comment:
This is used in only one place, so there's no need for a function. When reading the code, it'll make readers jump around to find out what this function is, only to discover that it's just a structure which isn't reused. Can we remove the function and just inline this?